src/Microsoft.IdentityModel.Tokens/Cache/MaximumSizeCache.cs

```diff
@@ -0,0 +1,272 @@
+
+        private const int CompactionStateNotRunning = 0; // no compaction running
+        private const int CompactionStateInProgress = 1; // compaction in progress
+        private int _compactionState = CompactionStateNotRunning;
+
+        #region constructors
+
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="cryptoProviderCacheOptions">Specifies the options which can be used to configure the cache settings.</param>
+        /// <param name="comparer">The equality comparison implementation to be used by the map when comparing keys.</param>
+        internal MaximumSizeCache(
+            CryptoProviderCacheOptions cryptoProviderCacheOptions,
+            IEqualityComparer<TKey> comparer = null)
+        {
+            _capacity = cryptoProviderCacheOptions.SizeLimit;
+            _compactionPercentage = cryptoProviderCacheOptions.CompactionPercentage;
+            _maxCapacityPercentage = cryptoProviderCacheOptions.MaxCapacityPercentage;
+            _map = new ConcurrentDictionary<TKey, CacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+        }
+
+        #endregion
+
+        #region public
+
+        /// <summary>
+        /// Gets or sets the value associated with the specified key.
+        /// </summary>
+        /// <param name="key">The key of the value to get or set.</param>
+        /// For get, if the specified key is not found, a <exception cref="KeyNotFoundException"> will be thrown.</exception>
+        /// For set, if the key is null throw an <exception cref="ArgumentNullException"> will be thrown.</exception>
+        public CacheItem<TKey, TValue> this[TKey key]
+        {
+            get => _map[key];
+            set => _map[key] = value;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <inheritdoc/>
+        public virtual void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
+        }
+
+        /// <inheritdoc/>
+        public virtual bool SetValue(TKey key, TValue value, DateTime expirationTime)
+        {
+            ValidateValues(key, value, expirationTime);
+            if (NeedsCompaction && Interlocked.CompareExchange(ref _compactionState, CompactionStateInProgress, CompactionStateNotRunning) == CompactionStateNotRunning)
+                ThreadPool.QueueUserWorkItem(CompactCache);
+
+            _map[key] = new CacheItem<TKey, TValue>(key, value, expirationTime);
+
+            return true;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            value = default;
```

Comment:
could you write this as:
```cs
return _map.TryGetValue(key, out CacheItem<TKey, TValue> value);

Reply:
The _map.TryGetValue(key, out CacheItem<TKey, TValue> value) is now wrapped in the TryGetCacheItem() which is shared by the derived class too, so it is more consistent.